
#Loan Default prediction using Machine Learning
##Predicting high- risk borrower with Logistic Regression, Random Forest, and SMOTE


#Business Problem/Project Overview
### Financial institutions need to identify borrowers who are likely to default on loans. this project uses machine learning to predict loan default risk and improve detection of high-risk customers

#Load data set

In [4]:
import pandas as pd
df = pd.read_csv('loan_data.csv')
df.head(7)

,credit.policy,purpose,int.rate,installment,log.annual.inc,dti,fico,days.with.cr.line,revol.bal,revol.util,inq.last.6mths,delinq.2yrs,pub.rec,not.fully.paid
0,1,debt_consolidation,0.1189,829.10,11.350407,19.48,737,5639.958333,28854,52.1,0,0,0,0
1,1,credit_card,0.1071,228.22,11.082143,14.29,707,2760.000000,33623,76.7,0,0,0,0
2,1,debt_consolidation,0.1357,366.86,10.373491,11.63,682,4710.000000,3511,25.6,1,0,0,0
3,1,debt_consolidation,0.1008,162.34,11.350407,8.10,712,2699.958333,33667,73.2,1,0,0,0
4,1,credit_card,0.1426,102.92,11.299732,14.97,667,4066.000000,4740,39.5,0,1,0,0
5,1,credit_card,0.0788,125.13,11.904968,16.98,727,6120.041667,50807,51.0,0,0,0,0
6,1,debt_consolidation,0.1496,194.02,10.714418,4.00,667,3180.041667,3839,76.8,0,0,1,1


# Understanding the dataset
To check the total rows, any missing values, datatypes,  and to confirm all columns.  

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9578 entries, 0 to 9577
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   credit.policy      9578 non-null   int64  
 1   purpose            9578 non-null   object 
 2   int.rate           9578 non-null   float64
 3   installment        9578 non-null   float64
 4   log.annual.inc     9578 non-null   float64
 5   dti                9578 non-null   float64
 6   fico               9578 non-null   int64  
 7   days.with.cr.line  9578 non-null   float64
 8   revol.bal          9578 non-null   int64  
 9   revol.util         9578 non-null   float64
 10  inq.last.6mths     9578 non-null   int64  
 11  delinq.2yrs        9578 non-null   int64  
 12  pub.rec            9578 non-null   int64  
 13  not.fully.paid     9578 non-null   int64  
dtypes: float64(6), int64(7), object(1)
memory usage: 1.0+ MB


The dataset contains 9,578 loan records with 14 features,including borrower financioal information and loan characteristics. The target variable, "not.fully.paid", indiucates whether a borrower defaulted on the loan. the data set is clean with no missing values, makiing it suitable for machine learning modeling

#check missing values and duplicate

In [6]:
df.isnull().sum()

,0
credit.policy,0
purpose,0
int.rate,0
installment,0
log.annual.inc,0
dti,0
fico,0
days.with.cr.line,0
revol.bal,0
revol.util,0


In [7]:
df.duplicated().sum()

np.int64(0)

#Identify target variable
##target variablke is 'not.fully.paid'
where: 0 = loan fully paid 1 = loan default

#Explore categorical column
##Categorical Feature Inspection
The 'purpose' column comtains different reasons for taking loans

In [8]:
df['purpose'].unique()

array(['debt_consolidation', 'credit_card', 'all_other',
       'home_improvement', 'small_business', 'major_purchase',
       'educational'], dtype=object)

The *purpose* column contains NorminalCategorical Values representing the reason for the loan using the .unique() function, seven distinct categories were identified, including debt consolidation, creditcard usage, and small business financing.

# Encode categorical variables

In [9]:
df = pd.get_dummies(df, columns=['purpose'], drop_first=True)
df.head()

,credit.policy,int.rate,installment,log.annual.inc,dti,fico,days.with.cr.line,revol.bal,revol.util,inq.last.6mths,delinq.2yrs,pub.rec,not.fully.paid,purpose_credit_card,purpose_debt_consolidation,purpose_educational,purpose_home_improvement,purpose_major_purchase,purpose_small_business
0,1,0.1189,829.10,11.350407,19.48,737,5639.958333,28854,52.1,0,0,0,0,False,True,False,False,False,False
1,1,0.1071,228.22,11.082143,14.29,707,2760.000000,33623,76.7,0,0,0,0,True,False,False,False,False,False
2,1,0.1357,366.86,10.373491,11.63,682,4710.000000,3511,25.6,1,0,0,0,False,True,False,False,False,False
3,1,0.1008,162.34,11.350407,8.10,712,2699.958333,33667,73.2,1,0,0,0,False,True,False,False,False,False
4,1,0.1426,102.92,11.299732,14.97,667,4066.000000,4740,39.5,0,1,0,0,True,False,False,False,False,False


The categorical variable purpose wasa transformed into using one-hot encoding, creating binary columns for each category. This allowed the machine learning model to interpret categorical information effectively

#Define features and Target


In [10]:
x = df.drop('not.fully.paid', axis=1)
y = df['not.fully.paid']

X- features(input) Y - target(what we predict)

#Split the Data

In [11]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y, test_size=0.2, random_state=42)

The dataset was split into training and testing sets using a 70-30 ratio. a random state was fixed to ensure reproducibility of results

In [12]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(x_train,y_train)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [13]:
y_pred = model.predict(x_test)


In [14]:
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_test,y_pred)
print(f"Accuracy: {accuracy * 100:.2f}%")

Accuracy: 84.24%


A logistic Regression model wwas trained to predict loan defaul. The model achieved an  accuracy of 84%%, with evaluation metrics such as precision, recall and f1 score used to assess performance

In [51]:
from sklearn.metrics import confusion_matrix
print(confusion_matrix(y_test, y_pred))

[[1609    2]
 [ 300    5]]


The defaulter matrix reaveals, that the model correctly identified 1,609 non defaulters but failed to detect most defaulters, correctly identifying only 5 out of 305. A large number of defaulters 300 were misclassified as safe borrowers, indicating poor performance in detecting high- risk customer.

#Classification Report

In [15]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.84      1.00      0.91      1611
           1       0.71      0.02      0.03       305

    accuracy                           0.84      1916
   macro avg       0.78      0.51      0.47      1916
weighted avg       0.82      0.84      0.77      1916



Although the model achieved an accuracy of 84.2%, it performed poorly in identifing loan defaulters, with a recall of only 2%. this indicates that the model fails to detect most high-risk customers, making it unstable for real- world financial decision making

#Random Forest Model

In [16]:
from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier()
rf_model.fit(x_train,y_train)


RandomForestClassifier()

o improve the model we imported RandomForestClassifier which is much stronger than Logistic Regression to handle imbalance better and captures complex pattern

In [20]:
y_pre_rf = rf_model.predict(x_test)

In [22]:
from sklearn.metrics import accuracy_score
print('Accuracy:',accuracy_score(y_test,y_pre_rf))
print(f"Accuracy: {accuracy * 100:.2f}%")

Accuracy: 0.8402922755741128
Accuracy: 84.24%


In [23]:
print(classification_report(y_test,y_pre_rf))

              precision    recall  f1-score   support

           0       0.84      0.99      0.91      1611
           1       0.47      0.03      0.05       305

    accuracy                           0.84      1916
   macro avg       0.66      0.51      0.48      1916
weighted avg       0.78      0.84      0.78      1916



The result shows that even after using Random Forest, the model predicts "no default" most of the time. It still fails to detect risky customers, with the recall for defaulters at just 3%. This suggests that the data is imbalanced and needs to be fix.

#Fix Imbalance

In [24]:
rf_model = RandomForestClassifier(class_weight ='balanced')
rf_model.fit(x_train,y_train)

RandomForestClassifier(class_weight='balanced')

The dataset was imbalanced , leading to poor detection of defaulters. To adress this, class weighting was applied to improve the model's ability to identify high-risk customers

In [52]:
y_pred_rf=rf_model.predict(x_test)

In [53]:
print(classification_report(y_test,y_pred_rf))

              precision    recall  f1-score   support

           0       0.87      0.84      0.86      1611
           1       0.30      0.36      0.33       305

    accuracy                           0.76      1916
   macro avg       0.59      0.60      0.59      1916
weighted avg       0.78      0.76      0.77      1916



The dataset was highly imbalanced, leading, leading to poor detection of defaulters. To adress this, SMOTE(SYthetic Minority Oversampling Technique)was applied to balance the classes. This significantly improved the model's ability to identify high-risk customers.

#Install and apply smote

In [54]:
!pip install imbalanced-learn

#Handle Class Imbalance with SMOTE

In [55]:
from imblearn.over_sampling import SMOTE
smote=SMOTE(random_state=42)
x_resampled,y_resampled=smote.fit_resample(x_train,y_train)


In [ ]:
rf_model = RandomForestClassifier()
rf_model.fit(x_resampled,y_resampled)

In [25]:
y_pred_smote = rf_model.predict(x_test)

In [26]:
print(classification_report(y_test,y_pred_smote))

              precision    recall  f1-score   support

           0       0.84      1.00      0.91      1611
           1       0.50      0.01      0.02       305

    accuracy                           0.84      1916
   macro avg       0.67      0.50      0.47      1916
weighted avg       0.79      0.84      0.77      1916



Initially , the logistc Regression and Random Forest models achieved high acccuracy but failed to identify loan defaulters due to class imbalance. After applying SMOTE, the recall for defaulters improved significantly from 2% to 24%, indicating better detection of high-risk customers. Although accuracy decreased, the model became more prectical for real-world financial risk assessment

# Model Tunning


In [56]:
from imblearn.over_sampling import SMOTE
smote=SMOTE(random_state=42)
x_resampled,y_resampled=smote.fit_resample(x_train,y_train)

In [57]:
rf_model = RandomForestClassifier(n_estimators=200,max_depth=10,random_state=42)
rf_model.fit(x_resampled,y_resampled)

RandomForestClassifier(max_depth=10, n_estimators=200, random_state=42)

In [30]:
y_ped_tuned = rf_model.predict(x_test)

In [34]:
from sklearn.metrics import accuracy_score
print('Accuracy:',accuracy_score(y_test,y_ped_tuned)*100)

Accuracy: 76.46137787056368


In [35]:
print(classification_report(y_test,y_ped_tuned))

              precision    recall  f1-score   support

           0       0.87      0.84      0.86      1611
           1       0.30      0.36      0.33       305

    accuracy                           0.76      1916
   macro avg       0.59      0.60      0.59      1916
weighted avg       0.78      0.76      0.77      1916



Aftger applying SMOTE and hyperparameter tunning , recall improved significantly to 36% demonstrating the model's improved ability to detect high-risk borrower The model performs very well in identifying non-defaulters(recall-84%) but strugggles to detect defaulters(recall = 30%). This indicates that while overall accuracy is high , the model has limited efectiveness in identifying high-risk customers

#Feature Importance

In [58]:
feature_importance = pd.Series(rf_model.feature_importances_, index=x_resampled.columns)
feature_importance.sort_values(ascending=False).head(10)

,0
credit.policy,0.205914
int.rate,0.117775
fico,0.096837
purpose_small_business,0.072107
purpose_debt_consolidation,0.059083
days.with.cr.line,0.053129
revol.bal,0.052079
log.annual.inc,0.049570
installment,0.049441
revol.util,0.049195


**Feature importance analysis** revealed that credit policy status, interest rate, and FICO score are the most influential factors in predicting loan default.

Borrowers who do not meet credit policy requirements and those with higher interest rates are significantly more likely to default.

Additionally, loan purpose plays a critical role, with small business and debt consolidation loans showing higher default risk.

Financial behavior indicators such as credit utilization, revolving balance, and income also contribute meaningfully to the model, highlighting the importance of borrower financial stability.

**The model aligns with real-world financial risk assessment principles, demonstrating that both institutional rules (credit policy) and individual financial behavior (income, utilization, credit history) jointly influence default risk.**

#Sample predictions

In [60]:
import pandas as pd

new_customer = pd.DataFrame({
    'credit.policy': [1],
    'int.rate': [0.15],
    'installment': [300],
    'log.annual.inc': [10.5],
    'dti': [18],
    'fico': [680],
    'days.with.cr.line': [4000],
    'revol.bal': [15000],
    'revol.util': [55],
    'inq.last.6mths': [1],
    'delinq.2yrs': [0],
    'pub.rec': [0],
    'not.fully.paid': [0],  # keep but not used
    # encoded columns (VERY IMPORTANT 👇)
    'purpose_credit_card': [0],
    'purpose_debt_consolidation': [1],
    'purpose_educational': [0],
    'purpose_home_improvement': [0],
    'purpose_major_purchase': [0],
    'purpose_small_business': [0]
})

In [62]:
new_customer = new_customer[x.columns]
prediction = rf_model.predict(new_customer)
print(prediction)

[1]


In [66]:
prob = rf_model.predict_proba(new_customer)
print(f"Probabilities: {prob[0][0]*100:.2f}% (Not Fully Paid), {prob[0][1]*100:.2f}% (Fully Paid)")

Probabilities: 48.31% (Not Fully Paid), 51.69% (Fully Paid)


The model predicts that the borrower is likely to repay the loan, wiyh a probability of 51.69%. However, the probability of default is also relatively high at 48.31%, indicating that this borrower represents a borderline risk case. Financial institutions should exercise caution when approving such loans

# Batch Mode

In [67]:
# Create a batch of applicants (you can expand this)
batch_applicants = pd.DataFrame({
    'credit.policy': [1, 1, 0, 1, 0],
    'int.rate': [0.12, 0.09, 0.15, 0.20, 0.07],
    'installment': [300, 200, 450, 500, 150],
    'log.annual.inc': [10.5, 11.2, 10.0, 9.8, 11.5],
    'dti': [15, 10, 25, 30, 8],
    'fico': [700, 750, 650, 600, 780],
    'days.with.cr.line': [4000, 5000, 3000, 2000, 6000],
    'revol.bal': [10000, 8000, 15000, 20000, 5000],
    'revol.util': [0.45, 0.30, 0.70, 0.85, 0.20],
    'inq.last.6mths': [1, 0, 2, 3, 0],
    'delinq.2yrs': [0, 0, 1, 2, 0],
    'pub.rec': [0, 0, 1, 1, 0]
})

# Get Probability Scores

In [68]:
processed_batch_applicants = batch_applicants.copy()
processed_batch_applicants['purpose'] = 'debt_consolidation'
processed_batch_applicants = pd.get_dummies(processed_batch_applicants, columns=['purpose'], drop_first=True)
model_features = x_resampled.columns
missing_cols = set(model_features) - set(processed_batch_applicants.columns)
for c in missing_cols:
    processed_batch_applicants[c] = 0
processed_batch_applicants = processed_batch_applicants[model_features]
probabilities = rf_model.predict_proba(processed_batch_applicants)[:, 1]
batch_applicants['default_probability'] = probabilities

# Create Risk Levels

In [69]:
def risk_level(prob):
    if prob < 0.30:
        return 'Low Risk'
    elif prob < 0.60:
        return 'Medium Risk'
    else:
        return 'High Risk'

batch_applicants['risk_level'] = batch_applicants['default_probability'].apply(risk_level)

# Add Business Decisions

In [70]:
def decision(risk):
    if risk == 'Low Risk':
        return 'Approve Loan'
    elif risk == 'Medium Risk':
        return 'Manual Review'
    else:
        return 'Reject / High Risk'

batch_applicants['decision'] = batch_applicants['risk_level'].apply(decision)

# Final Output

In [71]:
final_output = batch_applicants[
    ['default_probability', 'risk_level', 'decision']
].copy()

final_output['default_probability'] = final_output['default_probability'].round(2)

print(final_output)

   default_probability   risk_level       decision
0                 0.34  Medium Risk  Manual Review
1                 0.11     Low Risk   Approve Loan
2                 0.46  Medium Risk  Manual Review
3                 0.41  Medium Risk  Manual Review
4                 0.20     Low Risk   Approve Loan


In [ ]:
final_output.sort_values(by='default_probability', ascending=False)

# Business Deployment Simulation

This section demonstrates how the model can be used in a real-world financial setting. Instead of predicting one customer at a time, the model processes multiple loan applicants simultaneously, assigning each a probability of default.

Based on this probability, applicants are categorized into risk levels (Low, Medium, High) and assigned corresponding business actions such as loan approval, manual review, or rejection.

This approach enables financial institutions to automate risk assessment, prioritize high-risk applicants, and improve decision-making efficiency

## Key Findings

- Baseline models achieved high overall accuracy but failed to effectively identify defaulters, highlighting the impact of class imbalance on model performance.  
- The application of SMOTE significantly improved recall for defaulters (from near-zero to approximately 24–36%), enhancing the model’s ability to detect high-risk borrowers.  
- Credit policy status, interest rate, and FICO score emerged as the most influential predictors of loan default, aligning with real-world financial risk assessment principles.  
- The model demonstrated stronger performance as a risk detection tool, prioritizing the identification of high-risk applicants over overall accuracy.  
- When applied in a batch scoring framework, the model can evaluate multiple loan applicants simultaneously, assign default probabilities, and segment customers into actionable risk categories.  

## Final Project Summary

This project developed a machine learning-based credit risk assessment and decision-support system for predicting loan default.

While initial models struggled to detect defaulters due to class imbalance, the integration of SMOTE and model tuning significantly improved recall, making the model more effective for identifying high-risk borrowers.

Beyond prediction, the model was extended into a practical batch scoring system, capable of processing multiple loan applications at once. Each applicant is assigned a probability of default, categorized into risk levels (Low, Medium, High), and linked to business decisions such as approval, manual review, or rejection.

Feature importance analysis revealed key risk drivers—including credit policy, interest rate, and FICO score—providing actionable insights for improving lending strategies.

Overall, this project demonstrates how machine learning can be transformed from a predictive model into a scalable decision-support tool, enabling financial institutions to reduce default risk, improve operational efficiency, and make more informed lending decisions.